cdap-master/src/main/java/co/cask/cdap/data/tools/LogUpgrader.java

```diff
@@ -0,0 +1,203 @@
+
+import java.io.IOException;
+import java.net.URI;
+import java.net.URISyntaxException;
+import java.util.Map;
+
+/**
+ * Upgrades the Logs
+ */
+public class LogUpgrader extends AbstractUpgrader {
+
+  private static final Logger LOG = LoggerFactory.getLogger(LogUpgrader.class);
+  private static final String[] OTHERS = {"logs", Constants.SYSTEM_NAMESPACE, Constants.DEFAULT_NAMESPACE};
+
+  private final Transactional<DatasetContext<Table>, Table> logMDS;
+  private final CConfiguration cConf;
+  private final TransactionSystemClient txClient;
+  private final DatasetFramework nonNamespaedFramework;
+
+
+  @Inject
+  public LogUpgrader(LocationFactory locationFactory, final TransactionSystemClient txClient,
+                     final CConfiguration cConf,
+                     @Named("nonNamespacedDSFramework") final DatasetFramework nonNamespacedFramework) {
+    super(locationFactory);
+    this.cConf = cConf;
+    this.txClient = txClient;
+    this.nonNamespaedFramework = nonNamespacedFramework;
+    this.logMDS = Transactional.of(
+      new TransactionExecutorFactory() {
+        @Override
+        public TransactionExecutor createExecutor(Iterable<TransactionAware> txAwares) {
+          return new DefaultTransactionExecutor(txClient, txAwares);
+        }
+      },
+      new Supplier<DatasetContext<Table>>() {
+        @Override
+        public DatasetContext<Table> get() {
+          try {
+            return DatasetContext.of(new LogSaverTableUtil(nonNamespaedFramework, cConf).getMetaTable());
+          } catch (Exception e) {
+            throw Throwables.propagate(e);
+          }
+        }
+      });
+  }
+
+  @Override
+  public void upgrade() throws Exception {
+    final FileMetaDataManager fileMetaDataManager = new FileMetaDataManager(new LogSaverTableUtil(
+      nonNamespaedFramework, cConf), txClient, locationFactory);
+    logMDS.execute(new TransactionExecutor.Function<DatasetContext<Table>, Void>() {
+      @Override
+      public Void apply(DatasetContext<Table> ctx) throws Exception {
+        Scanner rows = ctx.get().scan(null, null);
+        Row row;
+        while ((row = rows.next()) != null) {
+          String key = Bytes.toString(row.getRow(), FileMetaDataManager.ROW_KEY_PREFIX.length,
+                                      (row.getRow().length - FileMetaDataManager.ROW_KEY_PREFIX.length));
+          Map<byte[], byte[]> allColumns = row.getColumns();
+          for (Map.Entry<byte[], byte[]> entry : allColumns.entrySet()) {
+            long startTimeMs = Bytes.toLong(entry.getKey());
+            String oldPath = Bytes.toString(entry.getValue());
+            Location newPath;
+            String newKey;
+            if (key.startsWith(Constants.Logging.SYSTEM_NAME) || key.startsWith(DEVELOPER_STRING)) {
+              newPath = upgradePath(key, oldPath);
+              newKey = upgradeKey(key);
+              try {
+                fileMetaDataManager.writeMetaData(newKey, startTimeMs, newPath);
```

Comment:
Should we delete the old entry?  What happens if we fail part way through. Next time it runs it will try to upgrade the same paths.  Is that handled correctly?